pkg/autodiscovery/providers/cloudfoundry.go

```diff
@@ -0,0 +1,206 @@
+		// `expandPerContainerChecks` if necessary
+		id := cloudfoundry.NewADNonContainerIdentifier(desiredLRP, adName)
+		// we need to convert adVal to map[string]string to pass it to extractTemplatesFromMap
+		convertedADVal := map[string]string{}
+		for k, v := range adVal {
+			convertedADVal[k] = string(v)
+		}
+		parsedConfigs, errs := extractTemplatesFromMap(id.String(), convertedADVal, "")
+		for _, err := range errs {
+			log.Errorf("Cannot parse endpoint template for service %s of app %s: %s, skipping",
+				adName, desiredLRP.AppGUID, err)
+		}
+
+		vcVal, vcOk := desiredLRP.EnvVcapServices[adName]
+		variables, varsOk := adVal["variables"]
+		success := false
+		switch {
+		case vcOk:
+			// if service is found in VCAP_SERVICES (non-container service), we will run a single check per App
+			err := cf.renderExtractedConfigs(parsedConfigs, variables, vcVal)
+			if err != nil {
+				log.Errorf("Failed to render config for service %s of app %s: %s", adName, desiredLRP.AppGUID, err)
+			} else {
+				success = true
+			}
+			cf.assignNodeNameToNonContainerChecks(parsedConfigs, desiredLRP, actualLRPs)
+		case varsOk:
+			log.Errorf("Service %s for app %s has variables configured, but is not present in VCAP_SERVICES", adName, desiredLRP.AppGUID)
+		default:
+			// if a service is not in VCAP_SERVICES and has no "variables" configured, we want to run a check per container
+			parsedConfigs = cf.expandPerContainerChecks(parsedConfigs, desiredLRP, actualLRPs, adName)
+			success = true
+		}
+
+		if success {
+			// mark all checks as cluster checks
+			for i := range parsedConfigs {
+				parsedConfigs[i].ClusterCheck = true
+			}
+			allConfigs = append(allConfigs, parsedConfigs...)
+		}
+	}
+
+	return allConfigs
+}
+
+func (cf CloudFoundryConfigProvider) assignNodeNameToNonContainerChecks(configs []integration.Config, desiredLRP cloudfoundry.DesiredLRP, actualLRPs []cloudfoundry.ActualLRP) {
+	if len(actualLRPs) > 0 {
+		aLRP := actualLRPs[0]
+		log.Debugf("All non-container checks for app %s will run on Cell %s", desiredLRP.AppGUID, aLRP.CellID)
+		for i := range configs {
+			configs[i].NodeName = aLRP.CellID
+		}
+	} else {
+		log.Infof("No container running for app %s, checks for its non-container services will run on arbitrary node", desiredLRP.AppGUID)
+	}
+}
+
+func (cf CloudFoundryConfigProvider) expandPerContainerChecks(
+	configs []integration.Config, desiredLRP cloudfoundry.DesiredLRP, actualLRPs []cloudfoundry.ActualLRP, svcName string) []integration.Config {
+	res := []integration.Config{}
+	for _, cfg := range configs {
+		for _, aLRP := range actualLRPs {
+			// we append container index to AD Identifier distinguish configs for different containers
+			newCfg := integration.Config{
+				ADIdentifiers: []string{cloudfoundry.NewADContainerIdentifier(desiredLRP, svcName, aLRP).String()},
+				ClusterCheck:  cfg.ClusterCheck,
+				Entity:        cfg.Entity, // TODO: should we modify this as well as ADIdentifiers?
```

Comment:
Let's either do this or track it somewhere else.

Reply:
Yeah, I'll do this, totally forgot to fix this TODO. Thanks for the catch!